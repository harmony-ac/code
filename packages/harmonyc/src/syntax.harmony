# Syntax

Empty file:
- empty => empty test design
- there are empty lines => empty test design


