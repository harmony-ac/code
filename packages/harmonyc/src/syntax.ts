import { Code, List, ListItem, RootContent as Node } from 'mdast'
import remarkParse from 'remark-parse'
import { unified } from 'unified'
import { Branch, Feature, Section, Step } from './model'
import {
  CucumberExpression,
  ParameterTypeRegistry,
} from '@cucumber/cucumber-expressions'

export interface ParsedFeature {
  name: string
  root: Section
}

const registry = new ParameterTypeRegistry()

export function parse({
  fileName,
  src,
}: {
  fileName: string
  src: string
}): Feature {
  const tree = unified().use(remarkParse).parse(src)
  const rootNodes = tree.children
  const feature = new Feature('')
  const headings = [feature.root]
  let name: string | undefined

  for (let i = 0; i < rootNodes.length; i++) {
    const node = rootNodes[i]
    if (node.type === 'heading') {
      if (node.depth === 1 && name === undefined) {
        name = textContent(node)
      }
      const level = node.depth
      const section = new Section(textContent(node), [], true)
      const last = headings.slice(0, level).reverse().find(Boolean)!
      last.addChild(section)
      headings[level] = section
      headings.length = level + 1
    } else {
      const last = headings[headings.length - 1]
      for (const branch of topLevel(node, feature)) last.addChild(branch)
    }
  }
  feature.name = name ?? fileName
  feature.root.setFeature(feature)
  return feature
}

function topLevel(node: Node, feature: Feature) {
  if (node.type === 'paragraph') return []
  if (node.type == 'list') return list(node)
  if (node.type === 'code') return code(node, feature)
  return []
}

function list(node: List) {
  const isFork = node.ordered === false
  return node.children.map((item) => listItem(item, isFork))
}

function code(node: Code, feature: Feature) {
  if (!node.meta?.match(/harmony/)) return []
  const code = node.value
  const marker = '///'
  const re = new RegExp(`^\s*${q(marker)}(.*?)$`, 'gm')
  let match = re.exec(node.value)
  const start = match?.index ?? code.length
  feature.prelude += code.slice(0, start)
  while (match) {
    const bodyStart = match.index + match[0].length
    const head = match[1].trim()
    match = re.exec(code)
    const end = match?.index ?? code.length
    const body = code.slice(bodyStart, end).trim()
    if (body) {
      feature.definitions.set(new CucumberExpression(head, registry), body)
    }
  }
  return []
}

function q(pattern: string) {
  return pattern.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

function listItem(node: ListItem, isFork: boolean) {
  const first = node.children[0]
  const text = textContent(first)
  let branch: Branch
  if (first.type === 'heading') {
    branch = new Section(text, [], isFork)
  } else {
    const [action, ...responses] = text.split(/(?:^| )=>(?: |$)/)
    branch = new Step(action, responses.filter(Boolean), [], isFork)
    const second = node.children[1]
    if (second?.type === 'code') {
      ;(branch as Step).action.docstring = second.value
    }
  }
  for (const child of node.children) {
    if (child.type === 'list') {
      for (const b of list(child)) branch.addChild(b)
    }
  }
  return branch
}

function textContent(node: Node): string {
  if (node.type === 'text') {
    return node.value.split(/\s+/).filter(Boolean).join(' ')
  }
  if (node.type === 'list') return ''
  if (!('children' in node)) return ''
  return node.children.map(textContent).filter(Boolean).join(' ')
}
