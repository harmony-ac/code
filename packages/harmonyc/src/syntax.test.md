# Syntax

Markdown code is mapped to the Harmony test model along the following rules.

## Empty file

- empty => empty test design
- there are empty lines => empty test design

  ```markdown

  ```

- there is only a paragraph => empty test design

  ```markdown
  The user should be able to log in.
  ```

## Section headings

- there is only a h1 => one section

  ```markdown
  # Login form
  ```

- there are more h1s => forked sections

  ```markdown
  # Login

  # Logout
  ```

- there are only h2s => forked sections

  ```markdown
  ## Login

  ## Logout
  ```

- there are nested headings => nested forked sections

  ```markdown
  # Login form

  ## Log in
  ```

- there are deep nested headings => deep nested forked sections

  ```markdown
  # Login form

  ## Log in

  ### Empty checks

  ### Email validation

  ## Log out

  ### Re-login
  ```

## Bulleted lists

- there is only a bullet point => one step

  ```markdown
  - log in
  ```

- there are more bullet points => forked steps
  ```markdown
  - log in
  - log out
  ```
- there are nested bullet points => nested forked steps
  ```markdown
  - log in
    - create order
    - log out
  - sign up
    - log out
  ```
- deep nested bullet points => deep nested forked steps

## Numbered lists

- there is only a numbered item => one step, not forked

  ```markdown
  1. log in
  ```

- there are more numbered items => sequential steps

  ```markdown
  1. log in
  2. log out
  ```

- there are nested bulleted and numbered lists => nested sequences and forks

  ```markdown
  - log in
    1. create order
       - pay
       - clear cart
    2. log out
  ```

- there are numbered and nested lists on same level => bulleted items are forks, numbered items not

  ```markdown
  - log in

  1. create order
  2. pay

  - log out

  1. log in
  2. log out
  ```

## Responses

Responses are specified with `=>` after the action.

- there is only a response => step with one response and empty action
  ```markdown
  - => logged out
  ```
- there is action and response => step with action and response
  ```markdown
  - log out => logged out
  ```
- there are more responses => step with more responses
  ```markdown
  - log out => logged out => on home page
  ```

## Headings within lists

- there is a heading in list item => one section
  ```markdown
  - # Login form
  ```
- there are headings in bulleted items => forked sections
  ```markdown
  - ### Login
  - ### Logout
  ```

## Docstrings

- there is a docstring => action with docstring

  ````markdown
  - the user writes an email

    ```
    Hello John,

    Please find the attached invoice.

    Best regards,
    Jane
    ```
  ````

<details><summary>Automation</summary>

```yaml harmony
automation:
  node:
    steps:
      empty: const input = ""
      preamble: |
        import expect from 'expect'
        import { parse } from './syntax'
      there is/are {}: const input = $_
      => {}: expect(parse(input)).toMatchSnapshot($1)
```

</details>
