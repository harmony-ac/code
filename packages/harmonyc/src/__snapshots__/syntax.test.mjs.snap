// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`T1 - Syntax - Empty file > empty test design 1`] = `[]`;

exports[`T2 - Syntax - Empty file > empty test design 1`] = `[]`;

exports[`T3 - Syntax - Empty file > empty test design 1`] = `[]`;

exports[`T4 - Syntax - Section headings > one section 1`] = `
[
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login form",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T5 - Syntax - Section headings > forked sections 1`] = `
[
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Logout",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T6 - Syntax - Section headings > nested forked sections 1`] = `
[
  Section {
    "children": [
      Section {
        "children": [],
        "isEnd": false,
        "isFork": true,
        "label": Label {
          "text": "Log in",
        },
        "parent": [Circular],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login form",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T7 - Syntax - Section headings > deep nested forked sections 1`] = `
[
  Section {
    "children": [
      Section {
        "children": [
          Section {
            "children": [],
            "isEnd": false,
            "isFork": true,
            "label": Label {
              "text": "Empty checks",
            },
            "parent": [Circular],
          },
          Section {
            "children": [],
            "isEnd": false,
            "isFork": true,
            "label": Label {
              "text": "Email validation",
            },
            "parent": [Circular],
          },
        ],
        "isEnd": false,
        "isFork": true,
        "label": Label {
          "text": "Log in",
        },
        "parent": [Circular],
      },
      Section {
        "children": [
          Section {
            "children": [],
            "isEnd": false,
            "isFork": true,
            "label": Label {
              "text": "Re-login",
            },
            "parent": [Circular],
          },
        ],
        "isEnd": false,
        "isFork": true,
        "label": Label {
          "text": "Log out",
        },
        "parent": [Circular],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login form",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T8 - Syntax - Bulleted lists > one step 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T9 - Syntax - Bulleted lists > forked steps 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T10 - Syntax - Bulleted lists > nested forked steps 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "create order",
        },
        "children": [],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "log out",
        },
        "children": [],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "sign up",
    },
    "children": [
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "log out",
        },
        "children": [],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T11 - Syntax - Bulleted lists > deep nested forked steps 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "create order",
        },
        "children": [
          Step {
            "action": Action {
              "feature": Feature {
                "definitions": Map {},
                "name": "test",
                "prelude": "",
                "root": Section {
                  "children": [Circular],
                  "isEnd": false,
                  "isFork": false,
                  "label": Label {
                    "text": "",
                  },
                },
              },
              "kind": "action",
              "text": "pay",
            },
            "children": [],
            "isEnd": false,
            "isFork": true,
            "parent": [Circular],
            "responses": [],
          },
          Step {
            "action": Action {
              "feature": Feature {
                "definitions": Map {},
                "name": "test",
                "prelude": "",
                "root": Section {
                  "children": [Circular],
                  "isEnd": false,
                  "isFork": false,
                  "label": Label {
                    "text": "",
                  },
                },
              },
              "kind": "action",
              "text": "clear cart",
            },
            "children": [],
            "isEnd": false,
            "isFork": true,
            "parent": [Circular],
            "responses": [],
          },
        ],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "log out",
        },
        "children": [],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "sign up",
    },
    "children": [
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "log out",
        },
        "children": [],
        "isEnd": false,
        "isFork": true,
        "parent": [Circular],
        "responses": [],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T12 - Syntax - Numbered lists > one step, not forked 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T13 - Syntax - Numbered lists > sequential steps 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T14 - Syntax - Responses > step with one response and empty action 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [
      Response {
        "feature": Feature {
          "definitions": Map {},
          "name": "test",
          "prelude": "",
          "root": Section {
            "children": [Circular],
            "isEnd": false,
            "isFork": false,
            "label": Label {
              "text": "",
            },
          },
        },
        "kind": "response",
        "text": "logged out",
      },
    ],
  },
]
`;

exports[`T15 - Syntax - Responses > step with more responses 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [
      Response {
        "feature": Feature {
          "definitions": Map {},
          "name": "test",
          "prelude": "",
          "root": Section {
            "children": [Circular],
            "isEnd": false,
            "isFork": false,
            "label": Label {
              "text": "",
            },
          },
        },
        "kind": "response",
        "text": "logged out",
      },
      Response {
        "feature": Feature {
          "definitions": Map {},
          "name": "test",
          "prelude": "",
          "root": Section {
            "children": [Circular],
            "isEnd": false,
            "isFork": false,
            "label": Label {
              "text": "",
            },
          },
        },
        "kind": "response",
        "text": "on home page",
      },
    ],
  },
]
`;

exports[`T16 - Syntax - Headings within lists > one section 1`] = `
[
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login form",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T17 - Syntax - Docstrings > action with docstring 1`] = `
[
  Step {
    "action": Action {
      "docstring": "Hello John,

Please find the attached invoice.

Best regards,
Jane",
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "the user writes an email",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T19 - Syntax - Section headings > forked sections 1`] = `
[
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Logout",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;

exports[`T20 - Syntax - Numbered lists > nested sequences and forks 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "create order",
        },
        "children": [
          Step {
            "action": Action {
              "feature": Feature {
                "definitions": Map {},
                "name": "test",
                "prelude": "",
                "root": Section {
                  "children": [Circular],
                  "isEnd": false,
                  "isFork": false,
                  "label": Label {
                    "text": "",
                  },
                },
              },
              "kind": "action",
              "text": "pay",
            },
            "children": [],
            "isEnd": false,
            "isFork": true,
            "parent": [Circular],
            "responses": [],
          },
          Step {
            "action": Action {
              "feature": Feature {
                "definitions": Map {},
                "name": "test",
                "prelude": "",
                "root": Section {
                  "children": [Circular],
                  "isEnd": false,
                  "isFork": false,
                  "label": Label {
                    "text": "",
                  },
                },
              },
              "kind": "action",
              "text": "clear cart",
            },
            "children": [],
            "isEnd": false,
            "isFork": true,
            "parent": [Circular],
            "responses": [],
          },
        ],
        "isEnd": false,
        "isFork": false,
        "parent": [Circular],
        "responses": [],
      },
      Step {
        "action": Action {
          "feature": Feature {
            "definitions": Map {},
            "name": "test",
            "prelude": "",
            "root": Section {
              "children": [Circular],
              "isEnd": false,
              "isFork": false,
              "label": Label {
                "text": "",
              },
            },
          },
          "kind": "action",
          "text": "log out",
        },
        "children": [],
        "isEnd": false,
        "isFork": false,
        "parent": [Circular],
        "responses": [],
      },
    ],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T21 - Syntax - Numbered lists > bulleted items are forks, numbered items not 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "create order",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "pay",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log in",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": false,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [],
  },
]
`;

exports[`T22 - Syntax - Responses > step with action and response 1`] = `
[
  Step {
    "action": Action {
      "feature": Feature {
        "definitions": Map {},
        "name": "test",
        "prelude": "",
        "root": Section {
          "children": [Circular],
          "isEnd": false,
          "isFork": false,
          "label": Label {
            "text": "",
          },
        },
      },
      "kind": "action",
      "text": "log out",
    },
    "children": [],
    "isEnd": false,
    "isFork": true,
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
    "responses": [
      Response {
        "feature": Feature {
          "definitions": Map {},
          "name": "test",
          "prelude": "",
          "root": Section {
            "children": [Circular],
            "isEnd": false,
            "isFork": false,
            "label": Label {
              "text": "",
            },
          },
        },
        "kind": "response",
        "text": "logged out",
      },
    ],
  },
]
`;

exports[`T23 - Syntax - Headings within lists > forked sections 1`] = `
[
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Login",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
  Section {
    "children": [],
    "isEnd": false,
    "isFork": true,
    "label": Label {
      "text": "Logout",
    },
    "parent": Section {
      "children": [Circular],
      "isEnd": false,
      "isFork": false,
      "label": Label {
        "text": "",
      },
    },
  },
]
`;
